File: kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # def treeSize(self, root: Optional[TreeNode]) -> int:
    #     if not root:
    #         return 0
    #     return 1 + self.treeSize(root.left) + self.treeSize(root.right)

    # def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
    #     index = self.treeSize(root.left) + 1
    #     print(index)
    #     if k > index:
    #         return self.kthSmallest(root.right, k - index)
    #     elif k < index:
    #         return self.kthSmallest(root.left, k)
    #     return root.val
    def inorder(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        lst = []
        lst.extend(self.inorder(root.left))
        lst.append(root.val)
        lst.extend(self.inorder(root.right))
        return lst
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:        
        return self.inorder(root)[k-1]
        
            

File: c++/kth-smallest-element-in-a-bst.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        vector<int> smallest;
        recursiveKthSmallest(root, k, smallest);
        return smallest[k - 1];
    }

    void recursiveKthSmallest(TreeNode* root, int k, vector<int>& smallest) {
        if (!k || root == nullptr) return;
        if (root->left == nullptr) {
            smallest.push_back(root->val);
            k--;
        } else {
            recursiveKthSmallest(root->left, k, smallest);
            if (k) {
                smallest.push_back(root->val);
                k--;
            }
        }
        recursiveKthSmallest(root->right, k, smallest);
    }
};